Sources/CombineWamp/Protocol/ElementType.swift

```diff
@@ -0,0 +1,178 @@
+            guard case let .integer(value) = self else { return nil }
+            return value
+        }
+        set {
+            guard case .integer = self, let newValue = newValue else { return }
+            self = .integer(newValue)
+        }
+    }
+
+    internal var string: String? {
+        get {
+            guard case let .string(value) = self else { return nil }
+            return value
+        }
+        set {
+            guard case .string = self, let newValue = newValue else { return }
+            self = .string(newValue)
+        }
+    }
+
+    internal var bool: Bool? {
+        get {
+            guard case let .bool(value) = self else { return nil }
+            return value
+        }
+        set {
+            guard case .bool = self, let newValue = newValue else { return }
+            self = .bool(newValue)
+        }
+    }
+
+    internal var double: Double? {
+        get {
+            guard case let .double(value) = self else { return nil }
+            return value
+        }
+        set {
+            guard case .double = self, let newValue = newValue else { return }
+            self = .double(newValue)
+        }
+    }
+
+    internal var dict: [String: ElementType]? {
+        get {
+            guard case let .dict(value) = self else { return nil }
+            return value
+        }
+        set {
+            guard case .dict = self, let newValue = newValue else { return }
+            self = .dict(newValue)
+        }
+    }
+
+    internal var list: [ElementType]? {
+        get {
+            guard case let .list(value) = self else { return nil }
+            return value
+        }
+        set {
+            guard case .list = self, let newValue = newValue else { return }
+            self = .list(newValue)
+        }
+    }
+}
+
+extension ElementType: Codable {
+    private struct AnyCodingKey: CodingKey {
+        var stringValue: String
+
+        init(stringValue: String) {
+            self.stringValue = stringValue
+        }
+
+        var intValue: Int?
+        init?(intValue: Int) {
+            return nil
+        }
+    }
+
+    internal init(from decoder: Decoder) throws {
+        if var listContainer = try? decoder.unkeyedContainer() {
+            var items: [ElementType] = []
+            while !listContainer.isAtEnd {
+                try items.append(listContainer.decode(ElementType.self))
+            }
+            self = .list(items)
+            return
+        }
+
+        if let dictContainer = try? decoder.container(keyedBy: AnyCodingKey.self) {
+            var items: [String: ElementType] = [:]
+            try dictContainer.allKeys.forEach {
+                items[$0.stringValue] = try dictContainer.decode(ElementType.self, forKey: $0)
+            }
+
+            self = .dict(items)
+            return
+        }
+
+        let singleValueContainer = try decoder.singleValueContainer()
+
+        if let integer = try? singleValueContainer.decode(Int.self) {
```

Comment:
oh noe, is this the famous trial and error decoding where the order is relevant? 😱

(not complaining about the code but about the spec)

Reply:
Unfortunately there's not much alternative here. Even for well known messages, such as HELLO, that we could theoretically decode based on a known schema, we still have the "Details"/"Options" dict (1) that could contain anything. And because there's not schema, and most of the times the data arrives in a [Any], all we can do is trial and error.

The order is relevant. This one seems to be working well, at least with the JSONs provided as example.

(1) https://wamp-proto.org/_static/gen/wamp_latest.html#extensibility